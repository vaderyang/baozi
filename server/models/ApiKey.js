// @flow
import { DataTypes, sequelize } from "../sequelize";
import randomstring from "randomstring";

const ApiKey = sequelize.define(
  "apiKeys",
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    name: DataTypes.STRING,
    secret: { type: DataTypes.STRING, unique: true },
    // TODO: remove this, as it's redundant with associate below
    userId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: {
        model: "users",
      },
    },
  },
  {
    tableName: "apiKeys",
    paranoid: true,
    hooks: {
      beforeValidate: key => {
        key.secret = randomstring.generate(38);
      },
    },
  }
);

ApiKey.associate = models => {
  ApiKey.belongsTo(models.User, {
    as: "user",
    foreignKey: "userId",
  });
};

export default ApiKey;
