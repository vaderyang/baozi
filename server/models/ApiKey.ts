import randomstring from "randomstring";
import { DataTypes, sequelize } from "../sequelize";

const ApiKey = sequelize.define(
  "apiKey",
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    name: DataTypes.STRING,
    secret: {
      type: DataTypes.STRING,
      unique: true,
    },
  },
  {
    paranoid: true,
    hooks: {
      // @ts-expect-error ts-migrate(7006) FIXME: Parameter 'key' implicitly has an 'any' type.
      beforeValidate: (key) => {
        key.secret = randomstring.generate(38);
      },
    },
  }
);

// @ts-expect-error ts-migrate(7006) FIXME: Parameter 'models' implicitly has an 'any' type.
ApiKey.associate = (models) => {
  ApiKey.belongsTo(models.User, {
    as: "user",
    foreignKey: "userId",
  });
};

export default ApiKey;
