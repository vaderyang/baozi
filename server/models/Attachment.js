// @flow
import path from "path";
import { DataTypes, sequelize } from "../sequelize";
import { deleteFromS3 } from "../utils/s3";

const Attachment = sequelize.define(
  "attachment",
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    key: {
      type: DataTypes.STRING,
      allowNull: false,
    },
    url: {
      type: DataTypes.STRING,
      allowNull: false,
    },
    contentType: {
      type: DataTypes.STRING,
      allowNull: false,
    },
    size: {
      type: DataTypes.BIGINT,
      allowNull: false,
    },
    acl: {
      type: DataTypes.STRING,
      allowNull: false,
      defaultValue: "public-read",
      validate: {
        isIn: [["private", "public-read"]],
      },
    },
  },
  {
    getterMethods: {
      name: function () {
        return path.parse(this.key).base;
      },
      redirectUrl: function () {
        return `/api/attachments.redirect?id=${this.id}`;
      },
      isPrivate: function () {
        return this.acl === "private";
      },
    },
  }
);

Attachment.beforeDestroy(async (model) => {
  await deleteFromS3(model.key);
});

Attachment.associate = (models) => {
  Attachment.belongsTo(models.Team);
  Attachment.belongsTo(models.Document);
  Attachment.belongsTo(models.User);
};

export default Attachment;
