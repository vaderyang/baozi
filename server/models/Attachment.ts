import path from "path";
import {
  BeforeDestroy,
  BelongsTo,
  Column,
  Default,
  ForeignKey,
  IsIn,
  Table,
  DataType,
  IsNumeric,
} from "sequelize-typescript";
import { publicS3Endpoint, deleteFromS3, getFileByKey } from "@server/utils/s3";
import Document from "./Document";
import Team from "./Team";
import User from "./User";
import IdModel from "./base/IdModel";
import Fix from "./decorators/Fix";
import Length from "./validators/Length";

@Table({ tableName: "attachments", modelName: "attachment" })
@Fix
class Attachment extends IdModel {
  @Length({
    max: 4096,
    msg: "key must be 4096 characters or less",
  })
  @Column
  key: string;

  @Length({
    max: 4096,
    msg: "url must be 4096 characters or less",
  })
  @Column
  url: string;

  @Length({
    max: 255,
    msg: "contentType must be 255 characters or less",
  })
  @Column
  contentType: string;

  @IsNumeric
  @Column(DataType.BIGINT)
  size: number;

  @Default("public-read")
  @IsIn([["private", "public-read"]])
  @Column
  acl: string;

  // getters

  get name() {
    return path.parse(this.key).base;
  }

  get redirectUrl() {
    return `/api/attachments.redirect?id=${this.id}`;
  }

  get isPrivate() {
    return this.acl === "private";
  }

  get buffer() {
    return getFileByKey(this.key);
  }

  /**
   * Use this instead of url which will be deleted soon, the column is unneccessary
   * and was not updated with the migraiton to the new s3 bucket.
   */
  get canonicalUrl() {
    return `${publicS3Endpoint()}/${this.key}`;
  }

  // hooks

  @BeforeDestroy
  static async deleteAttachmentFromS3(model: Attachment) {
    await deleteFromS3(model.key);
  }

  // associations

  @BelongsTo(() => Team, "teamId")
  team: Team;

  @ForeignKey(() => Team)
  @Column(DataType.UUID)
  teamId: string;

  @BelongsTo(() => Document, "documentId")
  document: Document;

  @ForeignKey(() => Document)
  @Column(DataType.UUID)
  documentId: string | null;

  @BelongsTo(() => User, "userId")
  user: User;

  @ForeignKey(() => User)
  @Column(DataType.UUID)
  userId: string;
}

export default Attachment;
