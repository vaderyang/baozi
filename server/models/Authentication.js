// @flow
import { DataTypes, sequelize, encryptedFields } from "../sequelize";

const Authentication = sequelize.define("authentication", {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
  },
  service: DataTypes.STRING,
  scopes: DataTypes.ARRAY(DataTypes.STRING),
  token: encryptedFields.vault("token"),
});

Authentication.associate = models => {
  Authentication.belongsTo(models.User, {
    as: "user",
    foreignKey: "userId",
  });
  Authentication.belongsTo(models.Team, {
    as: "team",
    foreignKey: "teamId",
  });
};

export default Authentication;
