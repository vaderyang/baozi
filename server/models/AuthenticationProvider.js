// @flow
import fs from "fs";
import path from "path";
import { DataTypes, sequelize } from "../sequelize";

// Each authentication provider must have a definition under server/auth, the
// name of the file will be used as reference in the db, one less thing to config
const authProviders = fs
  .readdirSync(path.resolve(__dirname, "..", "auth"))
  .filter(
    (file) =>
      file.indexOf(".") !== 0 &&
      !file.includes(".test") &&
      !file.includes("index.js")
  )
  .map((fileName) => fileName.replace(".js", ""));

const AuthenticationProvider = sequelize.define(
  "authentication_providers",
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    name: {
      type: DataTypes.STRING,
      validate: {
        isIn: [authProviders],
      },
    },
    enabled: {
      type: DataTypes.BOOLEAN,
      defaultValue: true,
    },
    providerId: {
      type: DataTypes.STRING,
    },
  },
  {
    timestamps: true,
    updatedAt: false,
  }
);

AuthenticationProvider.associate = (models) => {
  AuthenticationProvider.belongsTo(models.Team);
  AuthenticationProvider.hasMany(models.UserAuthentication);
};

export default AuthenticationProvider;
