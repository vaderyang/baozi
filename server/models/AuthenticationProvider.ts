import { ValidationError } from "../errors";
// @ts-expect-error ts-migrate(7034) FIXME: Variable 'providers' implicitly has type 'any[]' i... Remove this comment to see the full error message
import providers from "../routes/auth/providers";
import { DataTypes, Op, sequelize } from "../sequelize";

const AuthenticationProvider = sequelize.define(
  "authentication_providers",
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    name: {
      type: DataTypes.STRING,
      validate: {
        // @ts-expect-error ts-migrate(7005) FIXME: Variable 'providers' implicitly has an 'any[]' typ... Remove this comment to see the full error message
        isIn: [providers.map((p) => p.id)],
      },
    },
    enabled: {
      type: DataTypes.BOOLEAN,
      defaultValue: true,
    },
    providerId: {
      type: DataTypes.STRING,
    },
  },
  {
    timestamps: true,
    updatedAt: false,
  }
);

// @ts-expect-error ts-migrate(7006) FIXME: Parameter 'models' implicitly has an 'any' type.
AuthenticationProvider.associate = (models) => {
  AuthenticationProvider.belongsTo(models.Team);
  AuthenticationProvider.hasMany(models.UserAuthentication);
};

AuthenticationProvider.prototype.disable = async function () {
  const res = await AuthenticationProvider.findAndCountAll({
    where: {
      teamId: this.teamId,
      enabled: true,
      id: {
        [Op.ne]: this.id,
      },
    },
    limit: 1,
  });

  if (res.count >= 1) {
    return this.update({
      enabled: false,
    });
  } else {
    throw ValidationError("At least one authentication provider is required");
  }
};

AuthenticationProvider.prototype.enable = async function () {
  return this.update({
    enabled: true,
  });
};

export default AuthenticationProvider;
