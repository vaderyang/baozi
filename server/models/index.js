// @flow
import ApiKey from "./ApiKey";
import Attachment from "./Attachment";
import Authentication from "./Authentication";
import Backlink from "./Backlink";
import Collection from "./Collection";
import CollectionUser from "./CollectionUser";
import CollectionGroup from "./CollectionGroup";
import Document from "./Document";
import Event from "./Event";
import Integration from "./Integration";
import Group from "./Group";
import GroupUser from "./GroupUser";
import Notification from "./Notification";
import NotificationSetting from "./NotificationSetting";
import Revision from "./Revision";
import Share from "./Share";
import Star from "./Star";
import Team from "./Team";
import User from "./User";
import View from "./View";

const models = {
  ApiKey,
  Attachment,
  Authentication,
  Backlink,
  Collection,
  CollectionGroup,
  CollectionUser,
  Document,
  Event,
  Group,
  GroupUser,
  Integration,
  Notification,
  NotificationSetting,
  Revision,
  Share,
  Star,
  Team,
  User,
  View,
};

// based on https://github.com/sequelize/express-example/blob/master/models/index.js
Object.keys(models).forEach(modelName => {
  if ("associate" in models[modelName]) {
    models[modelName].associate(models);
  }
});

export {
  ApiKey,
  Attachment,
  Authentication,
  Backlink,
  Collection,
  CollectionGroup,
  CollectionUser,
  Document,
  Event,
  Group,
  GroupUser,
  Integration,
  Notification,
  NotificationSetting,
  Revision,
  Share,
  Star,
  Team,
  User,
  View,
};
