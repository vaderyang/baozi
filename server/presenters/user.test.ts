import { User } from "@server/models";
import presentUser from "./user";

it("presents a user", async () => {
  const user = presentUser(
    User.build({
      id: "123",
      name: "Test User",
      username: "testuser",
    })
  );
  expect(user).toMatchSnapshot();
});

it("presents a user without slack data", async () => {
  const user = presentUser(
    User.build({
      id: "123",
      name: "Test User",
      username: "testuser",
    })
  );
  expect(user).toMatchSnapshot();
});
