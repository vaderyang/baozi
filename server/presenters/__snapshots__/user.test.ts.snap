// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`presents a user 1`] = `
{
  "avatarUrl": null,
  "color": "#FF5C80",
  "createdAt": undefined,
  "id": "123",
  "isSuspended": false,
  "lastActiveAt": undefined,
  "name": "Test User",
  "role": "member",
  "updatedAt": undefined,
}
`;

exports[`presents a user without slack data 1`] = `
{
  "avatarUrl": null,
  "color": "#FF5C80",
  "createdAt": undefined,
  "id": "123",
  "isSuspended": false,
  "lastActiveAt": undefined,
  "name": "Test User",
  "role": "member",
  "updatedAt": undefined,
}
`;
